src/dune_pkg/opam_solver.ml

```diff
@@ -92,24 +86,9 @@ end = struct
     filter |> resolve_flags flags
     |> resolve_sys_var_treating_unset_vars_as_wildcards sys
 
-  let first_variable filter =
-    OpamFilter.fold_down_left
-      (fun acc filter ->
-        match (acc, filter) with
-        | None, FIdent (_, variable, _) -> Some variable
-        | _ -> acc)
-      None filter
-
   let eval_to_bool filter =
-    match filter with
-    | FIdent (_, variable, _) when is_variable_sys variable ->
-      Error (`Undefined_sys_env_var variable)
-    | _ -> (
-      match first_variable filter with
-      | Some variable -> Error (`Undefined_var variable)
-      | None -> (
-        try Ok (OpamFilter.eval_to_bool (Fun.const None) filter)
-        with Invalid_argument msg -> Error (`Not_a_bool msg)))
+    try Ok (OpamFilter.eval_to_bool ~default:false (Fun.const None) filter)
+    with Invalid_argument msg -> Error (`Not_a_bool msg)
```

Comment:
Just a note: it's pretty undesirable to have to rely on catching this exception. It can be thrown by just about any function in the Stdlib so it's not reliable to use it a signal for anything in particular.

Reply:
I agree. This is a limitation of the opam api. We can obviously change the api of our vendored copy of opam but I try to keep such changes to a minimum to simplify the eventual backporting effort.